src/core/lib/transport/handshaker.cc

```diff
@@ -101,6 +101,11 @@ bool HandshakeManager::CallNextHandshakerLocked(grpc_error_handle error) {
             HandshakerArgsString(&args_).c_str());
   }
   GPR_ASSERT(index_ <= handshakers_.size());
+  if (args_.read_buffer != nullptr) {
```

Comment:
We definitely do not want to do this here.  The read buffer has to be passed between handshakers and should not be destroyed unless handshaking fails.

Instead, I think we need to put this code back where it was, but allow it to be cleaned up if `read_buffer` is non-null, independently of whether the endpoint is null:

```
      if (args_.endpoint != nullptr) {
        // TODO(roth): It is currently necessary to shutdown endpoints
        // before destroying then, even when we know that there are no
        // pending read/write callbacks.  This should be fixed, at which
        // point this can be removed.
        grpc_endpoint_shutdown(args_.endpoint, error);
        grpc_endpoint_destroy(args_.endpoint);
      }
      if (args_.read_buffer != nullptr) {
        grpc_slice_buffer_destroy(args_.read_buffer);
        gpr_free(args_.read_buffer);
        args_.read_buffer = nullptr;
      }
      args_.endpoint = nullptr;
```

Reply:
(Made a small edit after sending)

Reply:
Done.